### PR TITLE
Support deffered types

### LINKED ISSUE #1: Resolve type hints from the container knowledge.
currently if you are not importing your dependencies the dep resolver wouldn't be able to find the
type your are referring to due to usage of `typing.get_type_hints` without namespace.
Is that intended or we can support resolving types based on the container knowledge?

### PR DESCRIPTION
**Pull request by nrbnlulu**
fix #1

### CODE REVIEW

**Review by ThirVondukr — CHANGES_REQUESTED** [review]
Thanks for the PR, I'd make a couple of changes, but everything looks good, I also tested it on my work codebase with around 200 dependencies and it works fine, though we didn't have to use deferred types, mostly because it'll probably mean that you'd have a circular dependency somewhere.

**Comment by nrbnlulu** [comment]
> mostly because it'll probably mean that you'd have a circular dependency somewhere.

I needed that due to `if TYPE_CHECKING` blocks.
If have some abstract class that have dependencies you usually won;t use them in the abstract class itself.
assuming that the implementations are in different modules.

File: aioinject/providers.py [hunk 1]
```diff
@@ -159,34 +166,59 @@ class DependencyLifetime(enum.Enum):
...
+        context: dict[str, Any] | None = None,
+    ) -> type[_T]:
+        ...
+
+    def resolve_type(self, context: dict[str, Any] | None = None) -> type[_T]:
+        if resolved := getattr(self, "_cached_type", None):
+            return resolved
+        ret = self._resolve_type_impl(context)
+        self._cached_type = ret
+        return ret
```
[the remark this answers is not in the document]
Comment (nrbnlulu) [reply aioinject/providers.py:191]: I think this might work though we'll need to initialise it on every instance ig

File: aioinject/providers.py [hunk 2]
```diff
@@ -159,34 +166,59 @@ class DependencyLifetime(enum.Enum):
 
 @runtime_checkable
 class Provider(Protocol[_T]):
-    type_: type[_T]
     impl: Any
     lifetime: DependencyLifetime
+    _cached_dependencies: tuple[Dependency[object], ...]
+    _cached_type: type[_T]  # TODO: I think it is redundant.
```
Comment (nrbnlulu) [inline aioinject/providers.py:172]: What do you think? I think this can be removed. I can find a valid usecase for this

Comment (nrbnlulu) [reply aioinject/providers.py:172]: You should provide the type in the provider I think no?

Comment (ThirVondukr) [reply aioinject/providers.py:172]: > You should provide the type in the provider I think no?

Not necessarily, in most cases type is derived from the object you pass into provider, e.g. `Scoped(Test)` would have type `Test`

**Comment by ThirVondukr** [comment]
I didn't include settings default `= None` for `_cached_dependencies` and `_cached_type`, but I think expecting AttributeError should work too, I think this can be merged

**Pull request merged**